packages/react-client/src/hooks/internal/device/useDevices.ts

```diff
@@ -0,0 +1,144 @@
+  const [audioError, setAudioError] = useState<DeviceError | null>(null);
+
+  const lastUsedCameraRef = useRef<MediaDeviceInfo | null>(persistHandlers?.getLastDevice("video") ?? null);
+  const lastUsedMicRef = useRef<MediaDeviceInfo | null>(persistHandlers?.getLastDevice("audio") ?? null);
+
+  const initializationRef = useRef<Promise<InitializeDevicesResult> | null>(null);
+
+  const saveUsedCamera = useCallback(
+    (deviceInfo: MediaDeviceInfo) => {
+      lastUsedCameraRef.current = deviceInfo;
+      persistHandlers?.saveLastDevice(deviceInfo, "video");
+    },
+    [persistHandlers],
+  );
+
+  const saveUsedMic = useCallback(
+    (deviceInfo: MediaDeviceInfo) => {
+      lastUsedMicRef.current = deviceInfo;
+      persistHandlers?.saveLastDevice(deviceInfo, "audio");
+    },
+    [persistHandlers],
+  );
+
+  const initializeDevices = useCallback(
+    async (settings?: { enableVideo?: boolean; enableAudio?: boolean }) => {
+      if (deviceList.length) {
+        return null;
+      }
+
+      const lastUsed = {
+        audio: lastUsedMicRef.current,
+        video: lastUsedCameraRef.current,
+      };
+
+      const constraints = {
+        video: settings?.enableVideo !== false && prepareConstraints(lastUsed.video?.deviceId, videoConstraints),
+        audio: settings?.enableAudio !== false && prepareConstraints(lastUsed.audio?.deviceId, audioConstraints),
+      };
+
+      const intitialize = async () => {
+        let result = await getAvailableMedia(constraints);
+        const devices = await navigator.mediaDevices.enumerateDevices();
+        setDeviceList(devices);
+
+        if (result.stream) {
+          result = await correctDevicesOnSafari(result.stream, result.errors, devices, constraints, lastUsed);
+        }
+
+        const videoDevice = deviceList.find(
+          (device) => device.deviceId === result.stream?.getVideoTracks()[0].getSettings().deviceId,
+        );
+        const audioDevice = deviceList.find(
+          (device) => device.deviceId === result.stream?.getAudioTracks()[0].getSettings().deviceId,
+        );
+
+        if (videoDevice) {
+          saveUsedCamera(videoDevice);
+        }
+        if (audioDevice) {
+          saveUsedMic(audioDevice);
+        }
+
+        setVideoStream(result.stream);
+        setAudioStream(result.stream);
+        setVideoError(result.errors.video);
+        setAudioError(result.errors.audio);
+
+        return result;
+      };
+
+      const initializePromise = intitialize();
+      initializationRef.current = initializePromise;
+
+      const result = await initializePromise;
+
+      initializationRef.current = null;
+      if (result.errors.video || result.errors.audio) return result.errors;
+      return null;
+    },
+    [deviceList, videoConstraints, audioConstraints, saveUsedCamera, saveUsedMic],
+  );
+
+  const getInitialStream = useCallback(async () => {
+    const result = await initializationRef.current;
+    return result?.stream ?? null;
```

Comment:
Is this expected that this method return non-null value only when `initializeDevices` is still running?